### PR TITLE
Adds Watchdog API and drivers for sifive,wdog0 and FE310-G000 LFROSC

### PR DESCRIPTION
**Pull request by nategraff-sifive**
Create a watchdog API and add a driver for sifive,wdog0 (sifive,aon0)

Also add a driver for FE310-G000 LFROSC.

### CODE REVIEW

File: src/drivers/sifive_fe310-g000_lfrosc.c [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+long __metal_driver_sifive_fe310_g000_lfrosc_get_rate_hz(const struct metal_clock *clock)
+{
+    struct metal_clock *internal_ref = __metal_driver_sifive_fe310_g000_lfrosc_lfrosc(clock);
+    struct metal_clock *external_ref = __metal_driver_sifive_fe310_g000_lfrosc_psdlfaltclk(clock);
+
+    unsigned long int cfg_reg = __metal_driver_sifive_fe310_g000_lfrosc_config_reg(clock);
+    unsigned long int mux_reg = __metal_driver_sifive_fe310_g000_lfrosc_mux_reg(clock);
+
+    if(LFROSC_REGW(mux_reg) & METAL_LFCLKMUX_EXT_MUX_STATUS) {
```
Comment (bsousi5) [inline src/drivers/sifive_fe310-g000_lfrosc.c]: Is there a configuration mean to know if there is an external clock feeding the mux, in addition to checking the clock?

File: src/drivers/sifive_wdog0.c [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+
+/* All writes to watchdog registers must be precedded by a write of
+ * a magic number to WDOGKEY */
+#define WDOG_UNLOCK(base) (WDOG_REGW(base, METAL_SIFIVE_WDOG0_WDOGKEY) = METAL_SIFIVE_WDOG0_MAGIC_KEY)
+
+int __metal_driver_sifive_wdog0_feed(const struct metal_watchdog *const wdog)
+{
+    const uintptr_t base = (uintptr_t)__metal_driver_sifive_wdog0_control_base(wdog);
+
+    WDOG_UNLOCK(base);
```
[the remark this answers is not in the document]
Comment (nategraff-sifive) [reply src/drivers/sifive_wdog0.c]: It's only necessary for writes though. `REGW` means "word (32-bit) sized register", not "write register".

File: src/drivers/sifive_fe310-g000_lfrosc.c (hunk 1 above)
Comment (nategraff-sifive) [reply src/drivers/sifive_fe310-g000_lfrosc.c]: Consider this for future improvement of the API.

File: src/drivers/sifive_wdog0.c [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+{
+    const uintptr_t base = (uintptr_t)__metal_driver_sifive_wdog0_control_base(wdog);
+    const struct metal_clock *const clock = __metal_driver_sifive_wdog0_clock(wdog);
+    
+    const long int clock_rate = metal_clock_get_rate_hz(clock);
+
+    if (clock_rate == 0)
+        return -1;
+
+    const unsigned int scale = (WDOG_REGW(base, METAL_SIFIVE_WDOG0_WDOGCFG) & METAL_WDOGCFG_SCALE_MASK);
```
[the remark this answers is not in the document]
Comment (nategraff-sifive) [reply src/drivers/sifive_wdog0.c]: No write here.

**Pull request merged**